File: neetcode/blind75/trees/level-order-traversal-of-binary-tree.py
# https://neetcode.io/problems/level-order-traversal-of-binary-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        res = []

        # BFS
        q = list()
        q.append(root)

        num_children = len(q)

        while len(q) > 0:
            sublist = list()
            for i in range(num_children):
                curr = q.pop(0)

                sublist.append(curr.val)

                if (curr.left):
                    q.append(curr.left)

                if (curr.right):
                    q.append(curr.right)

            res.append(sublist)
            num_children = len(q)

        return res

File: neetcode/blind75_JS/Trees/level-order-traversal-of-binary-tree.js
// https://neetcode.io/problems/level-order-traversal-of-binary-tree

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    levelOrder(root) {
        if (root === null){
            return []
        }

        const qu = [root]
        const res = []

        while (qu.length) {
            const subList = []
            let children = qu.length
            for (let i = 0; i < children; i ++) {
                let node = qu.shift()
                subList.push(node.val)

                if (node.left) {
                    qu.push(node.left)
                }

                if (node.right) {
                    qu.push(node.right)
                }
            }
            
            res.push(Array.from(subList))
            subList.splice(0)
        }
        return res
    }
}
